feat: add ability to delimit records within one COS object

dpittner opened the pull request:
This PR implements: https://github.com/ibm-messaging/kafka-connect-ibmcos-sink/issues/4 by allowing to specify an optional delimiter for records that are to be added into the same cos object.

Note that I hadn't had the chance to tests this with an actual Kafka.

mimaison commented on .gitignore:
```diff
@@ -2,5 +2,10 @@
+gradlew.bat
 bin/
 build/
+*.iml
```
nit: can we add a newline at the end of the file?

mimaison commented on src/main/java/com/ibm/eventstreams/connect/cossink/COSSinkConnectorConfig.java:
```diff
@@ -64,10 +64,16 @@
 
+    static final String CONFIG_NAME_COS_OBJECT_RECORD_DELIMITER = "cos.object.record.delimiter";
+    private static final String CONFIG_DOCUMENTATION_COS_OBJECT_RECORD_DELIMITER =
+            "The delimited to separate records within a single object storage object.";
```
typo: the "delimiter"

mimaison commented on src/main/java/com/ibm/eventstreams/connect/cossink/partitionwriter/COSObject.java:
```diff
@@ -41,8 +41,15 @@
+
+    COSObject(String recordDelimiter) {
+        if (recordDelimiter != null) {
+            LOG.trace("> setting recordDelimiter to: {} of length", recordDelimiter, recordDelimiter.length());
```
It's missing the 2nd `{}` after length

dpittner commented:
@mimaison comments should be addressed

mimaison commented:
@dpittner Thanks for the quick update! The change looks good.

Did you get a chance to test it? Or can @lijs-cn validate this works as expected?

mimaison commented:
@dpittner Can you also add the new configuration to the README?

lijs-cn commented:
Hi @dpittner ,

Thanks for providing the fix, I've verified the PR, and found when I specify `,` as delimiter set in `cos.object.record.delimiter`, it works. 

But when I want to set newline as delimiter, it is not as expected, no new lines are added. Could you please help to confirm whether I'm using the correct settings for newline delimiter?

My steps as below:
I've set `"cos.object.record.delimiter": "\n"` in cos-sink json configuration,  the configuration as below, and the object uploaded to COS is in attachment:
[kafka-java-console-sample-topic-1-0000000025904443-0000000025904447.txt](https://github.com/ibm-messaging/kafka-connect-ibmcos-sink/files/3528175/kafka-java-console-sample-topic-1-0000000025904443-0000000025904447.txt)


```
# curl http://localhost:8083/connectors/cos-sink-demo
{
    "name": "cos-sink-demo",
    "config": {
        "connector.class": "com.ibm.eventstreams.connect.cossink.COSSinkConnector",
        "cos.bucket.resiliency": "cross-region",
        "tasks.max": "1",
        "topics": "kafka-java-console-sample-topic",
        "cos.bucket.name": "****",
        "cos.api.key": "****",
        "cos.service.crn": "crn:v1:bluemix:public:cloud-object-storage:global:a/****:****::",
        "cos.object.record.delimiter": "\n",
        "name": "cos-sink-demo",
        "cos.endpoint.visibility": "public",
        "value.converter": "org.apache.kafka.connect.storage.StringConverter",
        "cos.bucket.location": "us",
        "key.converter": "org.apache.kafka.connect.storage.StringConverter",
        "cos.object.records": "5"
    },
    "tasks": [
        {
            "connector": "cos-sink-demo",
            "task": 0
        }
    ],
    "type": "sink"
}
```

dpittner commented:
I think the problem is with encoding a line break in a json with Java. As the only use case for now is to use a new line, I've adapted the code to be a boolean true/false that will include a new line directly or nothing (the default). I've also added a readme entry and adjusted the tests. the branch in my fork is updated. However I don't see the change. @lijs-cn can you give this another try? @mimaison should I open another PR or is there a way for me to update the code in this fork (e.g. pushing to my fork is pretty much the only thing I can do)

dpittner commented:
@lijs-cn please try using:  "cos.object.record.delimiter.nl": true with the new code, that should work regardless of config file encoding challenges

mimaison commented:
@dpittner If you push to your branch `dpittner:recordDelimiter` that will update this PR.

lijs-cn commented:
@dpittner I haven't found your code changes in this PR, so I verified using your personal clone directly, while sets ` "cos.object.record.delimiter.nl": true` in the connector config, it can generate COS object as expected, newlines are added for each event! I've also verified using IBM SQL Query to query from cos objects, works perfectly!

dpittner commented:
@mimaison I think something is odd, I did push to the branch in my fork, but no change here (e.g. that's also how @lijs-cn got the code). Likely I'm doing something wrong?

mimaison commented:
@dpittner Not sure what happened. I can see the new commit in your branch but that did not update the PR. Also I was able to find your commit in this repo but the branch info is missing: https://github.com/ibm-messaging/kafka-connect-ibmcos-sink/commit/ab9feec56a8857f199054211f06b062af5634595 :confused:

The best is probably to just open a new PR.

mimaison commented:
Ping @dpittner

dpittner commented:
@mimaison thanks for the ping, was on vacation (actually Bristol area ;-)) will open a new PR, thanks for following up!

dpittner commented:
done https://github.com/ibm-messaging/kafka-connect-ibmcos-sink/pull/6
